extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/HeaderCloudEventImpl.java

```diff
@@ -1,54 +1,91 @@
 package io.quarkus.funqy.runtime.bindings.knative.events;
 
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.reflect.Type;
 import java.time.OffsetDateTime;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+
+import io.quarkus.funqy.knative.events.AbstractCloudEvent;
 import io.quarkus.funqy.knative.events.CloudEvent;
-import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.MultiMap;
+import io.vertx.core.buffer.Buffer;
 
-class HeaderCloudEventImpl implements CloudEvent {
+class HeaderCloudEventImpl<T> extends AbstractCloudEvent<T> implements CloudEvent<T> {
     String id;
-    String specVersion;
+    CloudEvent.SpecVersion specVersion;
     String source;
+    String type;
+
     String subject;
     OffsetDateTime time;
+    Map<String, String> extensions;
+
+    String dataSchema;
+    String dataContentType;
+    T data;
 
-    final HttpServerRequest request;
+    final MultiMap headers;
+    final Buffer buffer;
+    final Type dataType;
+    final ObjectMapper mapper;
+    private ObjectReader reader;
 
-    HeaderCloudEventImpl(HttpServerRequest request) {
-        this.request = request;
+    HeaderCloudEventImpl(MultiMap headers, Buffer buffer, Type dataType, ObjectMapper mapper, ObjectReader reader) {
+        this.headers = headers;
+        this.buffer = buffer;
+        this.dataType = dataType;
+        this.mapper = mapper;
+        this.reader = reader;
     }
 
     @Override
     public String id() {
         if (id == null) {
-            id = this.request.getHeader("ce-id");
+            id = headers.get("ce-id");
         }
 
         return id;
     }
 
     @Override
-    public String specVersion() {
+    public SpecVersion specVersion() {
         if (specVersion == null) {
-            specVersion = this.request.getHeader("ce-specversion");
+            this.specVersion = SpecVersion.fromString(headers.get("ce-specversion"));
```

Comment:
I don't think spec version should be an enum.    Code can be deployed and ignored for a long time.  Dev apps should not have to upgrade, recompile, and redeploy their code just because the spec version string has changed from their cloud runtime.

Reply:
But if `specversion` changed then there might be incompatible changes to protocol.
For instance there might be a rename of a field. https://github.com/cloudevents/spec/issues/474
We really support only `0.3` and `1.0`.

Reply:
Bad specs aren't backward compatible.  I bet cloud events will remain backward compatible and will just add features, not remove them.

Reply:
Imagine having to upgrade and redeploy every single function in your system because the cloud environment now passes "1.0a" instead of "1.0" as a version.

Reply:
Agree that's not good. Well at least within major release there should be just compatible changes. So we can accept `1.SOMETHING`, but not `2.SOMETHING`.

Reply:
I don't think you should worry about the spec version so much.  Just default to highest version when creating an event if not set allow setting it as a string.